src/date-range-picker/use-date-range-picker.tsx

```diff
@@ -0,0 +1,106 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { useEffect, useState, useRef } from 'react';
+import { DateRangePickerProps, Focusable } from './interfaces';
+import { setTimeOffset } from './time-offset';
+
+const VALID_RANGE: DateRangePickerProps.ValidRangeResult = { valid: true };
+
+/**
+ * This function fills in a start and end time if they are missing.
+ */
+function fillMissingTime(value: DateRangePickerProps.AbsoluteValue | null) {
+  if (!value) {
+    return value;
+  }
+  const [startDate, startTime] = value.startDate.split('T');
+  const [endDate, endTime] = value.endDate.split('T');
+  return {
+    ...value,
+    startDate: startTime ? value.startDate : `${startDate}T00:00:00`,
+    endDate: endTime ? value.endDate : `${endDate}T23:59:59`,
+  };
+}
+
+function formatValue(
+  value: null | DateRangePickerProps.Value,
+  { timeOffset, dateOnly }: { timeOffset: number; dateOnly: boolean }
+): null | DateRangePickerProps.Value {
+  if (!value || value.type === 'relative') {
+    return value;
+  }
+  if (dateOnly) {
+    return {
+      type: 'absolute',
+      startDate: value.startDate.split('T')[0],
+      endDate: value.endDate.split('T')[0],
+    };
+  }
+  return setTimeOffset(value, timeOffset);
+}
+
+function getDefaultMode(
+  value: null | DateRangePickerProps.Value,
+  relativeOptions: readonly DateRangePickerProps.RelativeOption[],
+  rangeSelectorMode: DateRangePickerProps.RangeSelectorMode
+) {
+  if (value && value.type) {
+    return value.type;
+  }
+  if (rangeSelectorMode === 'relative-only') {
+    return 'relative';
+  }
+  if (rangeSelectorMode === 'absolute-only') {
+    return 'absolute';
+  }
+  return relativeOptions.length > 0 ? 'relative' : 'absolute';
+}
+
+export function useDateRangePicker({ value, relativeOptions, rangeSelectorMode }: UseDateRangePickerProps) {
+  const [rangeSelectionMode, setRangeSelectionMode] = useState<'absolute' | 'relative'>(
+    getDefaultMode(value, relativeOptions, rangeSelectorMode)
+  );
+
+  useEffect(() => {
+    setRangeSelectionMode(getDefaultMode(value, relativeOptions, rangeSelectorMode));
+  }, [rangeSelectorMode, value, relativeOptions]);
+
+  const [selectedAbsoluteRange, setSelectedAbsoluteRange] = useState<DateRangePickerProps.AbsoluteValue | null>(
+    value?.type === 'absolute' ? value : null
+  );
+
+  const [selectedRelativeRange, setSelectedRelativeRange] = useState<DateRangePickerProps.RelativeValue | null>(
+    value?.type === 'relative' ? value : null
+  );
+
+  const [validationResult, setValidationResult] = useState<
+    DateRangePickerProps.ValidRangeResult | DateRangePickerProps.InvalidRangeResult
+  >(VALID_RANGE);
+
+  const focusRefs = {
+    default: useRef<Focusable>(null),
+    'absolute-only': useRef<Focusable>(null),
+    'relative-only': useRef<Focusable>(null),
+  };
+
+  return {
+    VALID_RANGE,
```

Comment:
If it is a constant (never changes), why do you need to return it here?

UPD: same about other constant values. Why are they coming from hook and not `import {} from './utils'` for example?